db.go

```diff
@@ -0,0 +1,398 @@
+package netstats
+
+import (
+	"context"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+)
+
+type DB struct {
+	mu    sync.RWMutex
+	nodes map[string]*Node // items -- was array
+
+	blocks       map[string]*Block         // blocks by hash
+	propagations map[string][]*Propagation // propagations by hash
+
+	askedForHistory     bool
+	askedForHistoryTime time.Time
+
+	bestBlockNumberNotify chan struct{} // notify clients of new best block#
+
+	// Set of geolocations for trusted IPs. Must be set before using DB.
+	GeoByIP map[string]*Geo
+
+	// Now returns the current time, in milliseconds.
+	Now func() int
+}
+
+func NewDB() *DB {
+	return &DB{
+		nodes:        make(map[string]*Node),
+		blocks:       make(map[string]*Block),
+		propagations: make(map[string][]*Propagation),
+
+		bestBlockNumberNotify: make(chan struct{}),
+
+		GeoByIP: make(map[string]*Geo),
+
+		Now: func() int { return int(time.Now().UnixNano() / int64(time.Millisecond)) },
+	}
+}
+
+// BestBlockNumberNotify returns a channel that closes when a new best block number increases.
+func (db *DB) BestBlockNumberNotify() <-chan struct{} {
+	db.mu.RLock()
+	defer db.mu.RUnlock()
+	return db.bestBlockNumberNotify
+}
+
+// notifyBestBlockNumber closes the current notify channel and creates a new one.
+func (db *DB) notifyBestBlockNumber() {
+	close(db.bestBlockNumberNotify)
+	db.bestBlockNumberNotify = make(chan struct{})
+}
+
+// FindNodeByID returns a node by node ID.
+// Returns ErrNodeNotFound if node does not exist.
+func (db *DB) FindNodeByID(ctx context.Context, id string) (*Node, error) {
+	db.mu.RLock()
+	defer db.mu.RUnlock()
+
+	node := db.nodes[id]
+	if node == nil {
+		return nil, ErrNodeNotFound
+	}
+	return node.Clone(), nil
+}
+
+// Nodes returns a list of all nodes, sorted by id.
+func (db *DB) Nodes(ctx context.Context) ([]*Node, error) {
```

Comment:
Unused error?

Reply:
I left the `error` on here and `CreateNodeIfNotExists` because we may want to persist data or perform validation in the future and it seemed reasonable for `DB` calls to return errors generally.